Mathlib/Combinatorics/SetFamily/AhlswedeZhang.lean

```diff
@@ -0,0 +1,430 @@
+  · rw [←sum_congr rfl this, sum_range_sub', hf]
+    simp [choose_self, choose_zero_right, choose_eq_zero_of_lt h]
+  intro i h₁
+  rw [mem_range] at h₁
+  have h₁ := le_of_lt_succ h₁
+  have h₂ := h₁.trans_lt h
+  have h₃ := h₂.le
+  have hi₄ : (i + 1 : ℚ) ≠ 0 := by
+    have := (@cast_ne_zero ℚ _ _ _).mpr (succ_ne_zero i)
+    push_cast at this
+    exact this
+  have := congr_arg ((↑) : ℕ → ℚ) (choose_succ_right_eq m i)
+  push_cast at this
+  dsimp [hf]
+  rw [(eq_mul_inv_iff_mul_eq₀ hi₄).mpr this]
+  have := congr_arg ((↑) : ℕ → ℚ) (choose_succ_right_eq n i)
+  push_cast at this
+  rw [(eq_mul_inv_iff_mul_eq₀ hi₄).mpr this]
+  have : (m - i : ℚ) ≠ 0 := sub_ne_zero_of_ne (cast_lt.mpr h₂).ne'
+  have : (m.choose i : ℚ) ≠ 0 := cast_ne_zero.2 (choose_pos h₂.le).ne'
+  field_simp
+  ring
+
+private lemma Fintype.sum_div_mul_card_choose_card :
+    ∑ s : Finset α, (card α / ((card α - s.card) * (card α).choose s.card) : ℚ) =
+      card α * ∑ k in range (card α), (↑k)⁻¹ + 1 := by
+  rw [←powerset_univ, powerset_card_disjiUnion, sum_disjiUnion]
+  have : ∀ {x : ℕ}, ∀ s ∈ powersetCard x (univ : Finset α),
+    (card α / ((card α - Finset.card s) * (card α).choose (Finset.card s)) : ℚ) =
+      card α / ((card α - x) * (card α).choose x)
+  · intros n s hs
+    rw [mem_powersetCard_univ.1 hs]
+  simp_rw [sum_congr rfl this, sum_const, card_powersetCard, card_univ]
+  simp
+  simp_rw [mul_div, mul_comm, ←mul_div]
+  rw [←mul_sum, ←mul_inv_cancel (cast_ne_zero.mpr card_ne_zero : (card α : ℚ) ≠ 0), ←mul_add,
+    add_comm _ ((card α)⁻¹ : ℚ), ←
+    @sum_insert _ _ _ _ (fun x : ℕ ↦ (x⁻¹ : ℚ)) _ _ not_mem_range_self, ←range_succ]
+  have : ∀ x ∈ range (card α + 1),
+      ((card α).choose x / ((card α - x) * (card α).choose x) : ℚ) = (card α - x : ℚ)⁻¹
+  · intros n hn
+    rw [div_mul_left]
+    · simp
+    · exact cast_ne_zero.2 (choose_pos $ mem_range_succ_iff.1 hn).ne'
+  simp only [sum_congr rfl this, mul_eq_mul_left_iff, cast_eq_zero]
+  left
+  exact sum_bij (fun n _ ↦ card α - n) (fun a ha ↦ mem_range_succ_iff.mpr tsub_le_self)
+    (fun a ha ↦ by rw [cast_sub (mem_range_succ_iff.mp ha)])
+    (fun a₁ a₂ ha₁ ha₂ ↦
+      (tsub_right_inj (mem_range_succ_iff.mp ha₁) (mem_range_succ_iff.mp ha₂)).1)
+    fun b hb ↦ ⟨card α - b, mem_range_succ_iff.mpr tsub_le_self,
+      (tsub_tsub_cancel_of_le (mem_range_succ_iff.mp hb)).symm⟩
+
+end
+
+open scoped FinsetFamily
+
+namespace Finset
+variable {α β : Type*}
+
+/-! ### Truncated supremum, truncated infimum -/
+
+section SemilatticeSup
+variable [SemilatticeSup α] [OrderTop α] [@DecidableRel α (· ≤ ·)] [SemilatticeSup β]
+  [BoundedOrder β] [@DecidableRel β (· ≤ ·)] {s t : Finset α} {a b : α}
+
+private lemma sup_aux : a ∈ lowerClosure s → (s.filter fun b ↦ a ≤ b).Nonempty :=
+  fun ⟨b, hb, hab⟩ ↦ ⟨b, mem_filter.2 ⟨hb, hab⟩⟩
+
+/-- The infimum of the elements of `s` less than `a` if there are some, otherwise `⊤`. -/
```

Comment:
```suggestion
/-- The supremum of the elements of `s` less than `a` if there are some, otherwise `⊤`. -/
```
or else this needs some explanation of why the function is called `Sup`